src/query/execute_statement.c

```diff
@@ -14581,40 +14604,72 @@ do_prepare_subquery (PARSER_CONTEXT * parser, PT_NODE * stmt)
 {
   int err = NO_ERROR;
   PARSER_CONTEXT context;
-  int i, var_count;
-  PT_NODE *save_next = NULL;
+  int i, sub_idx, var_count;
+  PT_NODE *hv, *save_next = NULL;
+  PT_NODE **host_var_p;
   PT_MISC_TYPE save_flag;
 
   context = *parser;
 
-  context.host_var_count = 0;
   var_count = parser->host_var_count + parser->auto_param_count;
 
   stmt->info.query.flag.subquery_cached = 1;
 
+  context.host_var_count = 0;
+  context.host_variables = NULL;
+  stmt->sub_host_var_index = NULL;
+  stmt->sub_host_var_count = 0;
+
   if (var_count > 0)
     {
-      context.sub_host_var_index = (int *) parser_alloc (parser, var_count * sizeof (int));
-      if (context.sub_host_var_index == NULL)
+      host_var_p = (PT_NODE **) calloc (var_count, sizeof (PT_NODE *));
+      if (host_var_p == NULL)
+	{
+	  er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_OUT_OF_VIRTUAL_MEMORY, 1, var_count * sizeof (PT_NODE *));
+	  goto err_exit;
+	}
+
+      context.host_variables = (DB_VALUE *) malloc (var_count * sizeof (DB_VALUE));
+      if (context.host_variables == NULL)
+	{
+	  er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_OUT_OF_VIRTUAL_MEMORY, 1, var_count * sizeof (DB_VALUE));
+	  goto err_exit;
+	}
+
+      stmt->sub_host_var_index = (int *) parser_alloc (parser, var_count * sizeof (int));
+      if (stmt->sub_host_var_index == NULL)
 	{
 	  er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_OUT_OF_VIRTUAL_MEMORY, 1, var_count * sizeof (int));
-	  return ER_OUT_OF_VIRTUAL_MEMORY;
+	  goto err_exit;
 	}
-      /* to ininitialize empty index */
-      memset (context.sub_host_var_index, -1, var_count * sizeof (int));
-    }
 
-  /* not to traverse next subquery */
-  save_next = stmt->next;
-  stmt->next = NULL;
+      /* not to traverse next subquery */
+      save_next = stmt->next;
+      stmt->next = NULL;
 
-  parser_walk_tree (&context, stmt, pt_sub_host_vars_index, NULL, NULL, NULL);
+      parser_walk_tree (&context, stmt, pt_sub_host_vars_index, host_var_p, NULL, NULL);
 
-  /* restore next subquery */
-  stmt->next = save_next;
+      /* restore next subquery */
+      stmt->next = save_next;
+    }
+
+  for (i = 0; i < var_count; i++)
+    {
+      if (host_var_p[i])
+	{
+	  sub_idx = stmt->sub_host_var_count;
+	  stmt->sub_host_var_index[sub_idx] = i;
+	  db_value_clone (&parser->host_variables[i], &context.host_variables[sub_idx]);
+	  stmt->sub_host_var_count++;
+	  context.host_var_count++;
 
-  stmt->sub_host_var_count = context.host_var_count;
-  stmt->sub_host_var_index = context.sub_host_var_index;
+	  for (hv = host_var_p[i]; hv; hv = hv->info.host_var.next)
+	    {
+	      hv->info.host_var.saved = i;
```

Comment:
복구할 때 처리와 쌍을 이루도록 아래처럼 해 두는 것이  추후 코드 볼 때 이해하기가 쉬울 것 같습니다.
```suggestion
	      hv->info.host_var.saved = hv->info.host_var.index;
```

Reply:
네, 수정합니다.